src/mbgl/style/layer.cpp

```diff
@@ -106,5 +118,89 @@ optional<conversion::Error> Layer::setVisibility(const conversion::Convertible&
     return nullopt;
 }
 
+optional<std::string> LayerFactory::getSource(const conversion::Convertible& value) const {
+    auto sourceValue = objectMember(value, "source");
+    if (!sourceValue) {
+        return nullopt;
+    }
+
+    optional<std::string> source = toString(*sourceValue);
+    if (!source) {
+        return nullopt;
+    }
+
+    return source;
+}
+
+bool LayerFactory::initSourceLayerAndFilter(Layer* layer, const conversion::Convertible& value) const {
+    auto sourceLayerValue = objectMember(value, "source-layer");
+    if (sourceLayerValue) {
+        optional<std::string> sourceLayer = toString(*sourceLayerValue);
+        if (!sourceLayer) {
+            return false;
+        }
+        layer->setSourceLayer(*sourceLayer);
+    }
+
+    auto filterValue = objectMember(value, "filter");
+    if (filterValue) {
+        conversion::Error error;
+        optional<Filter> filter = conversion::convert<Filter>(*filterValue, error);
+        if (!filter) {
+            return false;
+        }
+        layer->setFilter(*filter);
+    }
+
+    return true;
+}
+
+// TODO: Move the LayerManager implementation to the dedicated .cpp file per platform.
+class LayerManagerImpl : public LayerManager {
+public:
+    void addLayerFactory(LayerFactory* factory) {
+        factories.emplace(std::make_pair(factory->type(), std::unique_ptr<LayerFactory>(factory)));
+    }
+private:
+    // LayerManager overrides.
+    std::unique_ptr<Layer> createLayer(const std::string& type, const std::string& id, const conversion::Convertible& value, conversion::Error& error) final;
+
+    std::unordered_map<std::string, std::unique_ptr<LayerFactory>> factories;
+};
+
+std::unique_ptr<Layer> LayerManagerImpl::createLayer(const std::string& type, const std::string& id, const conversion::Convertible& value, conversion::Error& error) {
+    auto search = factories.find(type);
+    if (search != factories.end()) {
+        if (auto layer = search->second->createLayer(id, value)) {
+            return layer;
+        }
+        error.message = "Error parsing a layer of type: " + type;
+    } else {
+        error.message = "Unsupported layer type: " + type;
+    }
+    return nullptr;
+}
+
+// static 
+LayerManager* LayerManager::get() {
+    static LayerManager* instance = nullptr;
+    if (instance == nullptr) {
+       static LayerManagerImpl impl;
+       impl.addLayerFactory(new FillLayerFactory);
```

Comment:
Great step towards making this modular.

Can we make this a `std::make_unique<XXXLayerFactory>()` to make the ownership transfer more explicit?

Reply:
forget about it, it would not work because it expects a `LayerFactory* factory` and having a `std::unique<LayerFactory>` won't help.